### PR TITLE
Implement STFT Decomposition transformer 

### PR DESCRIPTION
**Pull request by smk2007**
Implement STFT Decomposition transformer.

Certain hardware does not support DXIL, and therefore existing operator should be mapped to hardware supported functions.
Optimized convolution can be used to implement STFT.

### CODE REVIEW

File: onnxruntime/core/optimizer/stft_decomposition.cc [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+  if (std::is_same<T, float>::value) {
+    return ONNX_NAMESPACE::TensorProto_DataType_FLOAT;
+  } else if (std::is_same<T, MLFloat16>::value) {
+    return ONNX_NAMESPACE::TensorProto_DataType_FLOAT16;
+  } else if (std::is_same<T, double>::value) {
+    return ONNX_NAMESPACE::TensorProto_DataType_DOUBLE;
+  } else if (std::is_same<T, int64_t>::value) {
+    return ONNX_NAMESPACE::TensorProto_DataType_INT64;
+  }
+  return ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED;
```
Comment (PatriceVignola) [inline onnxruntime/core/optimizer/stft_decomposition.cc:34]: Should we throw an error instead if we somehow reach this line?

File: onnxruntime/core/optimizer/stft_decomposition.cc [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+
+STFTDecomposition::STFTDecomposition(const InlinedHashSet<std::string_view>& compatible_execution_providers) noexcept
+    : GraphTransformer("STFTDecomposition", compatible_execution_providers) {
+}
+
+template <typename T>
+constexpr static int32_t GetDataType() {
```
Comment (PatriceVignola) [inline onnxruntime/core/optimizer/stft_decomposition.cc:24]: Returning `TensorProto_DataType` would be more appropriate here.

File: onnxruntime/core/optimizer/stft_decomposition.cc [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+  }
+  return ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED;
+}
+
+template <typename TDataType, size_t TDims>
+NodeArg* AddInitializer(Graph& graph, const char* name, const int64_t (&shape)[TDims], const TDataType* begin) {
+  ONNX_NAMESPACE::TensorProto proto;
+  proto.set_name(graph.GenerateNodeArgName(name));
+  proto.set_data_type(GetDataType<TDataType>());
+  int64_t size = 1;
```
Comment (PatriceVignola) [inline onnxruntime/core/optimizer/stft_decomposition.cc:42]: I suggest using `element_count` or `num_elements` here (size is often used for the actual size in bytes of the tensor)

File: onnxruntime/core/optimizer/stft_decomposition.cc [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+std::pair<Node*, NodeArg*> AddNode(Graph& graph,
+                                   const char* op_type,
+                                   ProviderType execution_provider_type,
+                                   NodeArg* (&inputs)[TNumInputs]) {
+  auto def_name = graph.GenerateNodeArgName(op_type);
+  auto node_arg = &graph.GetOrCreateNodeArg(def_name, nullptr);
+  Node& node = graph.AddNode(graph.GenerateNodeName(op_type),
+                             op_type,
+                             "",
+                             gsl::make_span(inputs, inputs + TNumInputs),
```
Comment (PatriceVignola) [inline onnxruntime/core/optimizer/stft_decomposition.cc:67]: If you make `inputs` an `gsl::span` already, wouldn't you still be able to pass raw static arrays to this function and avoid making a span here? It feels like `gsl::span` should have an implicit constructor for it.

File: onnxruntime/core/optimizer/stft_decomposition.cc [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+      auto output_frame_length = stft.MutableOutputDefs()[0]->Shape()->dim(2).dim_value();
+      auto weight_size = static_cast<size_t>(dft_unique_bins * dft_size);
+      auto real_weights_data = std::vector<float>(weight_size);
+      auto imag_weights_data = std::vector<float>(weight_size);
+
+      // Populate weights
+      for (size_t k = 0; k < static_cast<size_t>(dft_unique_bins); k++) {
+        for (size_t n = 0; n < static_cast<size_t>(dft_size); n++) {
+          auto index = static_cast<size_t>(k * dft_size + n);
+          auto theta = -2 * 3.14159 * k * n / static_cast<float>(dft_size);
```
Comment (PatriceVignola) [inline onnxruntime/core/optimizer/stft_decomposition.cc:220]: Can we use `M_PI` here?

File: onnxruntime/core/providers/cpu/signal/dft.cc [hunk 6]
```diff
@@ -506,7 +506,7 @@ static Status short_time_fourier_transform(OpKernelContext* ctx, bool is_oneside
 
   // Calculate the window size with preference to the window input.
   const auto window_size = window ? window->Shape()[0] : frame_length;
-  ORT_ENFORCE(window_size < signal_size, "Ensure that the dft size is smaller than the signal.");
+  ORT_ENFORCE(window_size <= signal_size, "Ensure that the dft size is smaller than the signal.");
```
Comment (PatriceVignola) [inline onnxruntime/core/providers/cpu/signal/dft.cc:509]: `is not bigger than the signal`?

Comment (smk2007) [reply onnxruntime/core/providers/cpu/signal/dft.cc:509]: It is, but in theory could be equal. Saw this edge case, and decided to fix.

File: onnxruntime/core/optimizer/stft_decomposition.cc (hunk 1 above)
Comment (smk2007) [reply onnxruntime/core/optimizer/stft_decomposition.cc:34]: Replaced with static assert, and constexpr for ifs.

File: onnxruntime/core/optimizer/stft_decomposition.cc (hunk 3 above)
Comment (smk2007) [reply onnxruntime/core/optimizer/stft_decomposition.cc:42]: sounds good

File: onnxruntime/core/optimizer/stft_decomposition.cc (hunk 4 above)
Comment (smk2007) [reply onnxruntime/core/optimizer/stft_decomposition.cc:67]: yep, removed

File: onnxruntime/core/optimizer/stft_decomposition.cc (hunk 5 above)
Comment (smk2007) [reply onnxruntime/core/optimizer/stft_decomposition.cc:220]: Yep

**Review by PatriceVignola — DISMISSED** [review]
:shipit:

**Review by PatriceVignola — APPROVED** [review]

**Pull request merged**